### PR TITLE
[ffmpeg] Support splitting target based on extracted chapters (#24858)

### PR DESCRIPTION
**Pull request by wattux**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [ ] New extractor
- [x] New feature

---

### Description of your *pull request* and other information

As requested by [#24858](https://github.com/ytdl-org/youtube-dl/issues/24858), this adds the ability to split single tracks from a target file. Tracks are identified by the `chapters` list, provided from information extractors.

One thing remains unclear to me though:
I found it hard to create a universally valid filename. In `FFmpegSplitByTracksPP._build_track_name`, I did choose a rather naively approach by replacing `/` with `_`. But I'm sure there is already a util function for that, just could not find it.. 

Feel free to hit me with stuff to improve / change!

### CODE REVIEW

**Comment by unknownhellspawn** [comment]
FYI, I noticed a bug on this feature.  Youtube may pick up the chapters properly but if they have a mix of "hours" and "minutes", the preceeding track before it gets to the "hour" mark track seems to fail to split (not sure if that's an ffmpeg issue).

An example is this video:
https://www.youtube.com/watch?v=wY6_HQlCxHc

From Description:
0:00 Matoya’s Cave (FF1)
2:37 Theme of Love (FF4)
4:53 Main Theme of Final Fantasy 4 (Overworld)
8:12 Terra’s Theme (FF6)
10:44 Kids Run Through the City (FF6)
13:51 Phantom Forest (FF6)
16:56 Celes’s Theme (FF6)
19:42 Forever Rachel (FF6)
22:34 Aeris’s Theme (FF7)
26:13 Tifa’s Theme (FF7)
28:34 Main Theme of FF7 (Overworld/Cloud's Theme)
32:55 Ahead on our Way (FF7)
35:55 Balamb Garden (FF8)
38:56 Eyes on Me (FF8)
43:21 Crossing Those Hills (FF9)
47:32 Border Village Dali (FF9)
50:31 Terra (FF9)
54:58 Silence Before the Storm (FFX)
58:05 Eternity: Memory of Lightwaves (FFX-2)
1:01:44 Noctis (FFXV)

The track "Eternity" will generate a 0KB file in youtube-dl

Hope that helps ID the issue

**Comment by DrAutissimo** [comment]
The aforementioned issue is generally happening when a chapter's name contains ':', the created file will have the name of the chapter, without everything after and including the ':'.

**Comment by wattux** [comment]
@unknownhellspawn Thanks for the report! 

After rebasing on ytdl/master there was no problem with that link (anymore). E.g. "Eternity" had a reasonable file size, and the debug logs for ffmpeg `-ss` and `-t` arguments looked correctly. So I think the problem was an outdated master of youtube-dl

**Comment by DrAutissimo** [comment]
Issue still present, in this video, and any other where a chapter name contains a ':'
Using this command:
python .\youtube_dl\__main__.py -f bestvideo+bestaudio/best --download-archive archive.txt --merge-output-format MKV -i --all-subs --embed-subs --add-metadata --write-info-json --write-description -o "%(uploader)s/%(upload_date)s - %(title)s %(id)s.%(ext)s" --yes-playlist --split-tracks https://www.youtube.com/watch?v=wY6_HQlCxHc

As well as when using:
python .\youtube_dl\__main__.py -f bestvideo+bestaudio/best --split-tracks https://www.youtube.com/watch?v=wY6_HQlCxHc
or
python .\youtube_dl\__main__.py --split-tracks https://www.youtube.com/watch?v=wY6_HQlCxHc

It outputs all files correctly, except those where the chapter name contains ':', it is then cut, including all text after it, and no data is written to a file.

**Comment by DrAutissimo** [comment]
Issue still present, in this video, and any other where a chapter name contains a ':', I of course used the new push.
Using this command:
`python .\youtube_dl\__main__.py -f bestvideo+bestaudio/best --download-archive archive.txt --merge-output-format MKV -i --all-subs --embed-subs --add-metadata --write-info-json --write-description -o "%(uploader)s/%(upload_date)s - %(title)s %(id)s.%(ext)s" --yes-playlist --split-tracks https://www.youtube.com/watch?v=wY6_HQlCxHc`

As well as when using:
`python .\youtube_dl\__main__.py -f bestvideo+bestaudio/best --split-tracks https://www.youtube.com/watch?v=wY6_HQlCxHc`
or
`python .\youtube_dl\__main__.py --split-tracks https://www.youtube.com/watch?v=wY6_HQlCxHc
`
It outputs all files correctly, except those where the chapter name contains ':', it is then cut, including all text after it, and no data is written to a file.

**Comment by DrAutissimo** [comment]
I don't know what exactly broke now, but I retried this several times using this branch, but it now always gives me the 
`[ffmpeg] There are no tracks to extract`
No matter the video.

**Comment by DrAutissimo** [comment]
Yeah, this was related to how FFMPEG parses ':'.
I fixed it for myself by adding 

`track_title = track_title.replace(":"," ")`

This replaces the ':' with a space, obviously not ideal, but I don't know how to mark something as a char for FFMPEG.

**Comment by egrinant** [comment]
In Windows I had to implement additional rules for file names:
```
        track_title = track_title.replace(":", "_")
        track_title = track_title.replace("?", "_")
        track_title = track_title.replace("\\", "_")
        track_title = track_title.replace("*", "_")
        track_title = track_title.replace("'", "_")
        track_title = track_title.replace("\"", "_")
        track_title = track_title.replace("|", "_")
```
Also I've put an arbitrary length of 70 to the title, because of window's 260 character limitation in filenames.
```
track_title = (track_title[:70]) if len(track_title) > 70 else track_title
```

**Comment by DrAutissimo** [comment]
Yeah, that's essentially what I meant, although your fix is more comprehensive.

File: youtube_dl/options.py [hunk 1]
```diff
@@ -525,6 +525,11 @@ def _comma_separated_values_options_callback(option, opt_str, value, parser):
         '--external-downloader-args',
         dest='external_downloader_args', metavar='ARGS',
         help='Give these arguments to the external downloader')
+    downloader.add_option(
+        '--split-tracks',
+        dest='split_tracks', action='store_true', default=False,
+        help="Split tracks based on chapters information"
+    )
```
Comment (alkino) [inline youtube_dl/options.py:532]: I guess this option should be in `postproc` instead of `downloader`. isn't?

Comment (wattux) [reply youtube_dl/options.py:532]: Sounds reasonable, I'll have a look later

**Comment by alkino** [comment]
What about adding an option as `--output` to choose the names of files once splitted?

**Comment by femaref** [comment]
I've done some work based on the code in this PR, how would I go about adding that? do a new PR?

**Comment by wattux** [comment]
> I've done some work based on the code in this PR, how would I go about adding that? do a new PR?

Hey femaref, if you want to, I can invite you to my fork and we work on this further together. Otherwise, you'd have to create a new PR.

**Comment by femaref** [comment]
up to you - I'm fine with doing a new PR if you don't feel like working on this anymore (see https://github.com/femaref/youtube-dl/commits/split-track for what I've done so far).

**Comment by kynan** [comment]
There's conflicting information on whether or not embedding chapter markers in the video container is supported or not. Note: I *don't* want to split the video into multiple files.

#25356 suggests it "just works", but that's not my experience. The question wasn't really answered either and is now locked.

Couldn't find any relevant flags in the README / man page either, any clues?

**Comment by Autissima** [comment]
You don't need to add additional tags for that in my experience, it just works out of the box, the markers show up as chapters in VLC, with markings on the timeline, what build are you using and what video are you testing on?

**Comment by kynan** [comment]
I'm using 2021.02.10 (is that too old?) and used the video mentioned in #25356 to test.

**Comment by Autissima** [comment]
It should work, what command did you use?
I just confirmed that yes, it does work on my end with my general all purpose command of:
`.\youtube-dl.exe -v -f bestvideo+bestaudio/best --download-archive archive.txt --merge-output-format MKV -i --all-subs --embed-subs --add-metadata --write-info-json --write-description -o "%(uploader)s/%(upload_date)s - %(title)s %(id)s.%(ext)s" --yes-playlist`

**Comment by kynan** [comment]
Great, it is indeed the `--add-metadata` flag, thanks! :)

**Pull request closed without merging**